Orange/widgets/visualize/owlineplot.py

```diff
@@ -0,0 +1,762 @@
+
+        self.clear_messages()
+        if self.data is not None:
+            self.infoLabel.setText("%i instances on input\n%i features" % (
+                len(self.data), len(self.data.domain.attributes)))
+            self.graph_variables = [var for var in self.data.domain.attributes
+                                    if var.is_continuous]
+            self.valid_data = ~countnans(self.data.X, axis=1).astype(bool)
+            if len(self.graph_variables) < 1:
+                error(self.Error.not_enough_attrs)
+            elif not np.sum(self.valid_data):
+                error(self.Error.no_valid_data)
+            else:
+                if not np.all(self.valid_data):
+                    self.Information.hidden_instances()
+                if len(self.graph_variables) > MAX_FEATURES:
+                    self.Information.too_many_features()
+                    self.graph_variables = self.graph_variables[:MAX_FEATURES]
+
+    def check_display_options(self):
+        self.Warning.no_display_option.clear()
+        if self.data is not None:
+            if not (self.show_profiles or self.show_range or self.show_mean):
+                self.Warning.no_display_option()
+            enable = (self.show_profiles or self.show_range) and \
+                len(self.data[self.valid_data]) < SEL_MAX_INSTANCES
+            self.enable_selection.emit(enable)
+
+    @Inputs.data_subset
+    @check_sql_input
+    def set_subset_data(self, subset):
+        self.subset_data = subset
+
+    def handleNewSignals(self):
+        self.set_subset_ids()
+        if self.data is not None:
+            self._update_profiles_color()
+            self._update_sel_profiles_color()
+            self._update_sub_profiles()
+
+    def set_subset_ids(self):
+        sub_ids = {e.id for e in self.subset_data} \
+            if self.subset_data is not None else {}
+        self.subset_indices = None
+        if self.data is not None and sub_ids:
+            self.subset_indices = [x.id for x in self.data[self.valid_data]
+                                   if x.id in sub_ids]
+
+    def setup_plot(self):
+        if self.data is None:
+            return
+
+        ticks = [[(i + 1, str(a)) for i, a in enumerate(self.graph_variables)]]
+        self.graph.getAxis('bottom').setTicks(ticks)
+        self.plot_groups()
+        self.apply_selection()
+        self.graph.view_box.enableAutoRange()
+        self.graph.view_box.updateAutoRange()
+
+    def plot_groups(self):
+        self._remove_groups()
+        self.__groups = []
+        valid_indices = np.where(self.valid_data)[0]
+        data = self.data[self.valid_data, self.graph_variables]
+        if self.group_var is None:
+            self._plot_group(data, valid_indices)
+        else:
+            class_col_data, _ = self.data.get_column_view(self.group_var)
+            group_indices = [np.flatnonzero(class_col_data == i)
+                             for i in range(len(self.group_var.values))]
```

Comment:
On comment got lost in the previous review.

I think you can avoid a Python loop over all data instances by:

```
            class_col_data, _ = self.data.get_column_view(self.group_var)
            for index in range(len(self.group_var.values)):
                indices = np.flatnonzero(np.logical_and(class_col_data == index, self.valid_data))
                if not len(indices):
                    continue
                group_data = self.data[indices, self.graph_variables]
                self._plot_group(group_data, indices, index)
```

I haven't tried it (I ran a widget with this code, but I haven't experimented with invalid data).

If you decide to do it like this, you can also move `valid_indices` inside `if self.group_var is None` because it's no longer needed in the `else` part.

Reply:
👍